controllers/helmchart_controller.go

```diff
@@ -931,35 +936,79 @@ func (r *HelmChartReconciler) namespacedChartRepositoryCallback(ctx context.Cont
 		}
 		if secret, err := r.getHelmRepositorySecret(ctx, repo); secret != nil || err != nil {
 			if err != nil {
-				return nil, err
+				return nil, "", err
 			}
 			opts, err := getter.ClientOptionsFromSecret(*secret)
 			if err != nil {
-				return nil, err
+				return nil, "", err
 			}
 			clientOpts = append(clientOpts, opts...)
 
 			tlsConfig, err = getter.TLSClientConfigFromSecret(*secret, repo.Spec.URL)
 			if err != nil {
-				return nil, fmt.Errorf("failed to create TLS client config for HelmRepository '%s': %w", repo.Name, err)
+				return nil, "", fmt.Errorf("failed to create TLS client config for HelmRepository '%s': %w", repo.Name, err)
 			}
-		}
 
-		chartRepo, err := repository.NewChartRepository(repo.Spec.URL, "", r.Getters, tlsConfig, clientOpts)
-		if err != nil {
-			return nil, err
+			// Build registryClient options from secret
+			loginOpt, err := registry.LoginOptionFromSecret(repo.Spec.URL, *secret)
+			if err != nil {
+				return nil, "", fmt.Errorf("failed to create login options for HelmRepository '%s': %w", repo.Name, err)
+			}
+
+			loginOpts = append([]helmreg.LoginOption{}, loginOpt)
 		}
 
-		// Ensure that the cache key is the same as the artifact path
-		// otherwise don't enable caching. We don't want to cache indexes
-		// for repositories that are not reconciled by the source controller.
-		if repo.Status.Artifact != nil {
-			chartRepo.CachePath = r.Storage.LocalPath(*repo.GetArtifact())
-			chartRepo.SetMemCache(r.Storage.LocalPath(*repo.GetArtifact()), r.Cache, r.TTL, func(event string) {
-				r.IncCacheEvents(event, name, namespace)
-			})
+		var chartRepo chart.CleanDownloader
+		if helmreg.IsOCI(repo.Spec.URL) {
```

Comment:
Instead of if-else, I wonder if it'd be better and more consistent if we use switch statement based on `repo.spec.type`.
Since this helm repo is constructed above, if dependency resolution fails, we can set `repo.spec.type` value based on `IsOCI()` result. It'll just make the code look similar to `buildFromHelmRepository()` which may be a good thing.

Reply:
Yes switch would be more consistent and I will change this. I just don't think we will keep adding type or we will need to rethink the whole logic.

Reply:
I did not change to a switch statement finally because I am not setting the repository `type` field.